### PR TITLE
Add cross-compiling toolchain files for Android and Raspberry Pi.

### PR DESCRIPTION
**Pull request by Xreki**
- 增加了交叉编译Android、Raspberry Pi的cmake toolchain文件，自动地设置编译器、编译参数。
- 自动地编译host protobuf和target protobuf，OpenBLAS。

相关PR：#1698， #1835

### CODE REVIEW

File: cmake/cross_compiling/raspberry_pi.cmake [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+# This is a toolchain file for cross-compiling for Raspberry Pi.
+# The supported variables are listed belows:
+#
+# RPI_TOOLCHAIN
+# RPI_ARM_NEON
+#
+# Also you can set CMAKE_C/CXX_COMPILER yourself, through cmake arguments.
+
+IF(NOT RPI)
```
Comment (hedaoyuan) [inline cmake/cross_compiling/raspberry_pi.cmake]: 这个不需要判断吧，include(raspberry_pi.cmake)前已经判断了。

**Review by hedaoyuan — APPROVED** [review]
I can use `raspberry_pi.cmake` compiler raspberry pi version. But if use the `make j` command, the compilation process may be interrupted due to the error.

File: cmake/cross_compiling/raspberry_pi.cmake (hunk 1 above)
Comment (Xreki) [inline cmake/cross_compiling/raspberry_pi.cmake]: 每个toolchain file最开始都做了双重判断。留着保险一些吧。

**Comment by Xreki** [comment]
@hedaoyuan Thank you for your try. I have reproduced the compiling error. It is because of the misuse of `LIST(APPEND variable ...)`. Variables can not be modified in `function` using `LIST(APPEND variable ... `**PARENT_SCOPE**`)`  like `SET(variable ... `**PARENT_SCOPE**`)`. Thus the dependencies of `protobuf_host` and `protobuf` actually were not appended to `external_project_dependencies`. It is fixed in the latest commit and was tested using `make -j12`.

File: cmake/external/openblas.cmake [hunk 2]
```diff
@@ -21,36 +21,49 @@ IF(NOT ${CBLAS_FOUND})
...
-    ELSE(WIN32)
-        SET(CBLAS_LIBRARIES "${CBLAS_INSTALL_DIR}/lib/libopenblas.a" CACHE FILEPATH "openblas library" FORCE)
-    ENDIF(WIN32)
+    SET(CBLAS_LIBRARIES "${CBLAS_INSTALL_DIR}/lib/${LIBRARY_PREFIX}openblas${STATIC_LIBRARY_SUFFIX}"
+        CACHE FILEPATH "openblas library." FORCE)
+
+    SET(COMMON_ARGS CC=${CMAKE_C_COMPILER} NO_LAPACK=1 NO_SHARED=1)
+
+    IF(ANDROID)
+        SET(OPENBLAS_COMMIT "b5c96fcfcdc82945502a2303116a64d89985daf5")
```
Comment (gangliao) [inline cmake/external/openblas.cmake]: 用这个commit是什么原因呢

File: cmake/system.cmake [hunk 3]
```diff
@@ -72,12 +72,36 @@ MARK_AS_ADVANCED(HOST_SYSTEM CPU_CORES)
...
         SET(ANDROID TRUE)
+        INCLUDE(cross_compiling/android)
+    ELSEIF(${CMAKE_SYSTEM_NAME} STREQUAL "RPi")
+        SET(RPI TRUE)
+        INCLUDE(cross_compiling/raspberry_pi)
     ENDIF()
 ENDIF()
 
+# prefix and suffix on different os
+IF(WIN32)
```
Comment (gangliao) [inline cmake/system.cmake]: 👍

**Review by gangliao — APPROVED** [review]
LGTM

File: cmake/external/openblas.cmake (hunk 2 above)
Comment (Xreki) [inline cmake/external/openblas.cmake]: OpenBLAS为支持softfp新开的一个分支[arm_soft_fp_abi](https://github.com/xianyi/OpenBLAS/tree/arm_soft_fp_abi)，主分支只支持hardfp，Android上运行有些问题。

Comment (gangliao) [inline cmake/external/openblas.cmake]: 明白了

Comment (Xreki) [inline cmake/external/openblas.cmake]: 我加个注释说明一下吧~

**Comment by Xreki** [comment]
@hedaoyuan 建议写一个`如何编写交叉编译toolchain文件`的设计文档，该文档将和`如何交叉编译ARM版Paddle`的使用文档一起，在下一个PR中提交。

**Pull request merged**